cmd/oras/discover.go

```diff
@@ -0,0 +1,207 @@
+/*
+Copyright The ORAS Authors.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"os"
+	"strings"
+
+	"oras.land/oras-go/v2/registry/remote"
+	"oras.land/oras/cmd/oras/internal/option"
+
+	"github.com/need-being/go-tree"
+	digest "github.com/opencontainers/go-digest"
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+	artifactspec "github.com/oras-project/artifacts-spec/specs-go/v1"
+	"github.com/spf13/cobra"
+)
+
+type discoverOptions struct {
+	option.Common
+	option.Remote
+
+	targetRef    string
+	artifactType string
+	outputType   string
+}
+
+func discoverCmd() *cobra.Command {
+	var opts discoverOptions
+	cmd := &cobra.Command{
+		Hidden: true,
+		Use:    "discover [options] <name:tag|name@digest>",
+		Short:  "Discover referrers of a manifest in the remote registry",
+		Long: `Discover referrers of a manifest in the remote registry
+
+Example - Discover direct referrers of manifest 'hello:latest' in registry 'localhost:5000':
+  oras discover localhost:5000/hello
+
+Example - Discover all the referrers of manifest 'hello:latest' in registry 'localhost:5000' in a tree view:
+  oras discover localhost:5000/hello -o tree
+
+Example - Discover referrers with type 'test-artifact' of manifest 'hello:latest' in registry 'localhost:5000':
+  oras discover --artifact test-artifact localhost:5000/hello
+`,
+		Args: cobra.ExactArgs(1),
+		PreRun: func(cmd *cobra.Command, args []string) {
+			fmt.Println("Command discover is in preview and might have breaking changes coming.")
+		},
+		RunE: func(cmd *cobra.Command, args []string) error {
+			opts.targetRef = args[0]
+			return runDiscover(opts)
+		},
+	}
+
+	cmd.Flags().StringVarP(&opts.artifactType, "artifact-type", "", "", "artifact type")
+	cmd.Flags().StringVarP(&opts.outputType, "output", "o", "table", "format in which to display referrers (table, json, or tree). tree format will also show indirect referrers")
+	option.ApplyFlags(&opts, cmd.Flags())
+	return cmd
+}
+
+func runDiscover(opts discoverOptions) error {
+	ctx, _ := opts.SetLoggerLevel()
+	repo, err := opts.NewRepository(opts.targetRef, opts.Common)
+	if err != nil {
+		return err
+	}
+
+	// discover artifacts
+	root := tree.New(opts.targetRef)
+	desc, err := repo.Resolve(ctx, repo.Reference.ReferenceOrDefault())
```

Comment:
@qweeah @Wwwsylvia @FeynmanZhou @yizha1 We need to make the default behavior consistent for all commands.

`latest` tag by default is convenient but also causes lots of problems.

At least, we need to inform the user that we are using a default tag like docker does.
```
$ docker pull registry
Using default tag: latest
...
```

This is achievable by
```go
ref := repo.Reference.ReferenceOrDefault()
if ref != repo.Reference.Reference {
    fmt.Println("Using default tag:", ref)
}
```

Reply:
Let me update this way but I think we do need a proper discussion on that.